// @bugsnag/react-native-performance/dist/auto-instrumentation/app-start-plugin.js
import React from 'react';
import { createAppStartSpan } from '../create-app-start-span.js';

const isWrapperComponentProvider = (value) => value === null || typeof value === 'function';
class AppStartPlugin {
    constructor(appStartTime, spanFactory, clock, appRegistry) {
        this.appStartTime = appStartTime;
        this.spanFactory = spanFactory;
        this.clock = clock;
        this.appRegistry = appRegistry;
    }
    configure(configuration) {
        if (!configuration.autoInstrumentAppStarts)
            return;
        const appStartSpan = createAppStartSpan(this.spanFactory, this.appStartTime);
        const AppStartWrapper = ({ children }) => {
            React.useEffect(() => {
                if (appStartSpan.isValid()) {
                    this.spanFactory.endSpan(appStartSpan, this.clock.now());
                }
            }, []);
            return children;
        };
        const instrumentedComponentProvider = (appParams) => ({ children }) => {
            if (configuration.wrapperComponentProvider) {
                const WrapperComponent = configuration.wrapperComponentProvider(appParams);
                return (<AppStartWrapper>
            <WrapperComponent>{children}</WrapperComponent>
          </AppStartWrapper>);
            }
            return <AppStartWrapper>{children}</AppStartWrapper>;
        };
        this.appRegistry.setWrapperComponentProvider(instrumentedComponentProvider);
    }
}

export { AppStartPlugin, isWrapperComponentProvider };
