// @rohitrehan/react-native-country-picker-modal/lib/CloseButton.js
import React from 'react';
import { Image, Platform, StyleSheet, TouchableNativeFeedback, TouchableOpacity, View, } from 'react-native';
import useTheme from './CountryTheme';
import closeImage from '../assets/images/close.android.png';
import closeImageIOS from '../assets/images/close.ios.png';
const styles = StyleSheet.create({
    container: {
        height: 48,
        width: '15%',
        alignItems: 'center',
        justifyContent: 'center',
    },
    imageStyle: {
        height: 25,
        width: 25,
        resizeMode: 'contain',
    },
});
const CloseButtonAndroid = (props) => {
    const { onBackgroundTextColor } = useTheme();
    return (<View style={[styles.container, props.style]}>
      <TouchableNativeFeedback background={typeof Platform.Version === 'number' && Platform.Version < 21
            ? TouchableNativeFeedback.SelectableBackground()
            : TouchableNativeFeedback.SelectableBackgroundBorderless()} onPress={props.onPress}>
        <View>
          <Image source={props.image ?? closeImage} style={[
            styles.imageStyle,
            props.imageStyle,
            { tintColor: onBackgroundTextColor },
        ]}/>
        </View>
      </TouchableNativeFeedback>
    </View>);
};
const CloseButtonIOS = (props) => {
    const { onBackgroundTextColor } = useTheme();
    return (<View style={[styles.container, props.style]}>
      <TouchableOpacity onPress={props.onPress}>
        <Image source={props.image ?? closeImageIOS} style={[
            styles.imageStyle,
            props.imageStyle,
            { tintColor: onBackgroundTextColor },
        ]}/>
      </TouchableOpacity>
    </View>);
};
export default Platform.select({
    ios: CloseButtonIOS,
    android: CloseButtonAndroid,
    web: CloseButtonIOS,
    default: CloseButtonIOS,
});
