// @rohitrehan/react-native-country-picker-modal/lib/index.js
import CountryModal from './CountryModal';
import { DARK_THEME } from './CountryTheme';
import CountryFilter from './CountryFilter';
import CountryList from './CountryList';
import FlagButton from './FlagButton';
import Flag from './Flag';
import HeaderModal from './HeaderModal';
import CountryModalProvider from './CountryModalProvider';
import React from 'react';
import CountryContext, { DEFAULT_COUNTRY_CONTEXT } from './CountryContext';
import { ThemeProvider, DEFAULT_THEME } from './CountryTheme';
import CountryPicker from './CountryPicker';
const Main = ({ theme, translation, ...props }) => {
    return (<ThemeProvider theme={{ ...DEFAULT_THEME, ...theme }}>
      <CountryContext.Provider value={{ ...DEFAULT_COUNTRY_CONTEXT, translation }}>
        <CountryPicker {...props}/>
      </CountryContext.Provider>
    </ThemeProvider>);
};
export default Main;
export { getCountriesAsync as getAllCountries, getCountryCallingCodeAsync as getCallingCode, } from './CountryService';
export * from './types';
export { CountryFilter, CountryList, FlagButton, Flag, CountryModal, HeaderModal, CountryModalProvider, DARK_THEME, DEFAULT_THEME, CountryPicker, };
