// @bugsnag/react-native-performance/dist/config.js
import { schema, isStringWithLength, isBoolean, isStringOrRegExpArray } from '@bugsnag/core-performance';
import { defaultNetworkRequestCallback, isNetworkRequestCallback } from '@bugsnag/request-tracker-performance';
import { isWrapperComponentProvider } from './auto-instrumentation/app-start-plugin.js';

function createSchema() {
    return {
        ...schema,
        codeBundleId: {
            defaultValue: '',
            message: 'should be a string',
            validate: isStringWithLength
        },
        generateAnonymousId: {
            defaultValue: true,
            message: 'should be true|false',
            validate: isBoolean
        },
        autoInstrumentAppStarts: {
            defaultValue: true,
            message: 'should be true|false',
            validate: isBoolean
        },
        wrapperComponentProvider: {
            defaultValue: null,
            message: 'should be a function',
            validate: isWrapperComponentProvider
        },
        autoInstrumentNetworkRequests: {
            defaultValue: true,
            message: 'should be true|false',
            validate: isBoolean
        },
        networkRequestCallback: {
            defaultValue: defaultNetworkRequestCallback,
            message: 'should be a function',
            validate: isNetworkRequestCallback
        },
        tracePropagationUrls: {
            defaultValue: [],
            message: 'should be an array of string|RegExp',
            validate: isStringOrRegExpArray
        }
    };
}

export { createSchema as default };
